src/slurm_plugin/slurm_resources/slurm_resources.py

```diff
@@ -0,0 +1,413 @@
+# Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+# with the License. A copy of the License is located at
+#
+# http://aws.amazon.com/apache2.0/
+#
+# or in the "LICENSE.txt" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+# OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions and
+# limitations under the License.
+
+import collections
+import logging
+from abc import abstractmethod
+from enum import Enum
+
+from common.utils import time_is_up
+
+logger = logging.getLogger(__name__)
+
+# Possible ec2 health status: 'ok'|'impaired'|'insufficient-data'|'not-applicable'|'initializing'
+EC2_HEALTH_STATUS_UNHEALTHY_STATES = {"impaired"}
+# Possible instance states: 'pending'|'running'|'shutting-down'|'terminated'|'stopping'|'stopped'
+EC2_INSTANCE_HEALTHY_STATES = {"pending", "running"}
+EC2_INSTANCE_STOP_STATES = {"stopping", "stopped"}
+EC2_INSTANCE_ALIVE_STATES = EC2_INSTANCE_HEALTHY_STATES | EC2_INSTANCE_STOP_STATES
+EC2_SCHEDULED_EVENT_CODES = [
+    "instance-reboot",
+    "system-reboot",
+    "system-maintenance",
+    "instance-retirement",
+    "instance-stop",
+]
+
+
+NodeType = collections.namedtuple("NodeType", ["instance_type", "partition"])
+EC2Instance = collections.namedtuple("EC2Instance", ["id", "private_ip", "hostname", "launch_time"])
+CONFIG_FILE_DIR = "/etc/parallelcluster/slurm_plugin"
+
+
+class PartitionStatus(Enum):
+    UP = "UP"
+    DOWN = "DOWN"
+    INACTIVE = "INACTIVE"
+    DRAIN = "DRAIN"
+
+    def __str__(self):
+        return str(self.value)
+
+
+class SlurmPartition:
+    def __init__(self, name, nodenames, state):
+        """Initialize slurm node with attributes."""
+        self.name = name
+        self.nodenames = nodenames
+        self.state = state
+        self.slurm_nodes = []
+
+    def is_active(self):
```

Comment:
Maybe we should do `is_inactive` instead, because partitions can be in other states that we might want to use in the future